lib/pavilion/plugins/commands/graph.py

```diff
@@ -0,0 +1,313 @@
+
+                label = evaluations[evl].split(".")[0]
+                label = label if label not in labels else ""
+                labels.add(label)
+
+                x_list = [x] * len(y_list)
+
+                try:
+                    matplotlib.pyplot.scatter(x=x_list, y=y_list, marker=".",
+                                              color=color,
+                                              label=label)
+                except ValueError:
+                    output.fprint("Evaluations '{}, {}' resulted in "
+                                  "un-plottable point with types ({}, {})."
+                                  .format(evaluations['x'], evaluations[evl],
+                                          type(x_list[-1]).__name__,
+                                          type(y_list[-1]).__name__))
+
+                    return errno.EINVAL
+
+                if evaluations[evl] in args.plot_average:
+                    stats_dict[evl]['x'].append(x)
+                    stats_dict[evl]['y'].append(statistics.mean(y_list))
+
+        for evl, values in stats_dict.items():
+            if evaluations[evl] not in args.plot_average:
+                continue
+            label = evaluations[evl].split(".")[0]
+            matplotlib.pyplot.scatter(values['x'], values['y'],
+                                      marker="+",
+                                      color=colormap[evl]['stat'],
+                                      label="average({})".format(label))
+
+        matplotlib.pyplot.ylabel(args.ylabel)
+        matplotlib.pyplot.xlabel(args.xlabel)
+        matplotlib.pyplot.legend()
+
+        matplotlib.pyplot.savefig(args.filename)
+        output.fprint("Completed. Graph saved as '{}.png'."
+                      .format(args.filename), color=output.GREEN,
+                      file=self.outfile)
+
+    def gather_results(self, evaluations, test_results) -> Dict:
+        """
+        Gather and format a test run objects results.
+        :param evaluations: The evaluations dictionary to be used to gather
+               results.
+        :param test_results: A test run's result dictionary.
+        :return: result, a dictionary containing parsed/formatted results for
+                 the given test run.
+        """
+
+        evaluate_results(test_results, evaluations)
+
+        # X value evaluations should only result in a list when graphing
+        # individual node results from a single test run.
+        if isinstance(test_results['x'], list):
+            if not isinstance(test_results['x'][-1], (int, str, float)):
+                raise ResultTypeError("x value evaluation '{}' resulted in a "
+                                      "list of invalid type {}."
+                                      .format(evaluations['x'],
+                                              type(test_results['x'][-1])
+                                              .__name__))
+            results = {}
```

Comment:
Give a comment about what data structure you're building, so it's clear what is trying to do.